### DISCUSSION TITLE
mrgsolve error compiling models

### DISCUSSION THREAD

**Discussion by soutomas** [metrumresearchgroup/mrgsolve#799] [Q&A]
I've been using mrgsolve for a while now. I've recently updated R to v4.0.2 with rtools40 and the latest mrgsolve on CRAN, and started seeing this error when compiling with this code that ran without any problems before that update. Does anyone know what's wrong?

> Building m23 ... Error in file(con, "r") : cannot open the connection

**Comment by kylebaron** [comment]
Can you compile anything?  Or just `m23`?

**Comment by soutomas** [comment]
I have a few models to compile, m23 is the frist one. I tried skipping m23 as a test but the others have the same error.

**Comment by kylebaron** [comment]
Ok; I'd look at the Rtools40 installation; were you able to install a package from source?  See the example below:


![Screen Shot 2020-09-12 at 7 38 13 PM](https://user-images.githubusercontent.com/16908301/93007555-9dec5280-f52f-11ea-8e2d-893856ea23b6.png)

**Comment by soutomas** [comment]
Yes, I followed the examples below from the rtools site and no problem installing from the source.

**Comment by kylebaron** [comment]
Sounds good. Happy to keep working with you on this.  Let me know if there are any other leads to investigate.

**Comment by soutomas** [comment]
Thanks Kyle. Since that's the only error message I got, I'm not sure what else I can look into at this point.

**Comment by kylebaron** [comment]
You could share more information: 

- what does the output look like when you compiled the package from source?
- can you debug `mread` and let us know at what point you get the error (`?debugonce`)?
- did you blow away everything from your R 3.x install? Could your system be looking at / for old files?
- what is the `sessionInfo()`?
- what do you get with `pkgbuild::check_build_tools()`?

Etc ... there must be more information that could be shared that might suggest what is happening on your system.

**Comment by kylebaron** [comment]
- Is this a network drive? 
- Are you working with a network drive at all? 
- Are there spaces in the path to R?
- Does this work: `tmp <- tempfile(); cat(letters,file = tmp); readLines(tmp)` (can you write to `tempdir()`)?'
- This: `pkgbuild::check_compiler(debug = TRUE)`?

**Comment by kylebaron** [comment]
@tmscs Any luck with any of this?

**Comment by soutomas** [comment]
Thanks for the tips Kyle. Apology for my delayed response. I've now managed to test your suggestions as below. 

#### what does the output look like when you compiled the package from source?
This is the output when I installed the testing json package on the rtool40 website:
> :/rtools40/mingw64/bin/gcc -shared -s -static-libgcc -o jsonlite.dll tmp.def base64.o collapse_array.o collapse_object.o collapse_pretty.o escape_chars.o integer64_to_na.o is_datelist.o is_recordlist.o is_scalarlist.o modp_numtoa.o null_to_na.o num_to_char.o parse.o prettify.o push_parser.o r-base64.o register.o row_collapse.o transpose_list.o validate.o -Lyajl -lstatyajl -LC:/PROGRA~1/R/R-4.0.2/bin/x64 -lR
installing to C:/Program Files/R/R-4.0.2/library/00LOCK-jsonlite/00new/jsonlite/libs/x64
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
  converting help for package 'jsonlite'
    finding HTML links ... done
    base64                                  html  
    flatten                                 html  
    fromJSON                                html  
    prettify                                html  
    rbind_pages                             html  
    read_json                               html  
    serializeJSON                           html  
    stream_in                               html  
    unbox                                   html  
    validate                                html  
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
*** arch - i386
*** arch - x64
** testing if installed package can be loaded from final location
*** arch - i386
*** arch - x64
** testing if installed package keeps a record of temporary installation path
**  DONE (jsonlite)
>
> The downloaded source packages are in
        ‘C:\Users\más\AppData\Local\Temp\RtmpMdwROc\downloaded_packages’
> 

#### can you debug mread and let us know at what point you get the error (?debugonce)?
The is the output when I tried `debugonce`. I've omitted the model code since it's long and probably not the issue. 

> debugging in: mcode_cache("m23", s23code)
debug: {
    mread_cache(model, project, code = code, ...)
}
Browse[2]> model
[1] "m23"
Browse[2]> project
[1] "C:\\Users\\Tomás\\AppData\\Local\\Temp\\RtmpmGKSPx"
Browse[2]> code
[1] "\n$PARAM @annotated\......."
Browse[2]> 
debug: mread_cache(model, project, code = code, ...)
Browse[2]> 
Building m23 ... Error in file(con, "r") : cannot open the connection
> 
> 



#### did you blow away everything from your R 3.x install? Could your system be looking at / for old files?
I think so. 

#### what is the sessionInfo()?
I've checked this before and couldn't detect any irregularities:
> R version 4.0.2 (2020-06-22)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19041)
>
>Matrix products: default
>
>locale:
[1] LC_COLLATE=English_Australia.1252 
[2] LC_CTYPE=English_Australia.1252   
[3] LC_MONETARY=English_Australia.1252
[4] LC_NUMERIC=C                      
[5] LC_TIME=English_Australia.1252    
>
>atached base packages:
[1] stats     graphics  grDevices utils     datasets  methods  
[7] base     
>
>other attached packages:
[1] ggpubr_0.4.0    scales_1.1.1    ggplot2_3.3.2   mrgsolve_0.10.4
[5] magrittr_1.5    tidyr_1.1.1     dplyr_1.0.1     plyr_1.8.6     
>
>loaded via a namespace (and not attached):
 [1] zip_2.1.0                 Rcpp_1.0.5               
 [3] cellranger_1.1.0          pillar_1.4.6             
 [5] compiler_4.0.2            forcats_0.5.0            
 [7] tools_4.0.2               lifecycle_0.2.0          
 [9] tibble_3.0.3              gtable_0.3.0             
[11] pkgconfig_2.0.3           rlang_0.4.7              
[13] openxlsx_4.1.5            curl_4.3                 
[15] haven_2.3.1               rio_0.5.16               
[17] RcppArmadillo_0.9.900.2.0 withr_2.2.0              
[19] generics_0.0.2            vctrs_0.3.2              
[21] hms_0.5.3                 grid_4.0.2               
[23] tidyselect_1.1.0          glue_1.4.1               
[25] data.table_1.13.0         R6_2.4.1                 
[27] rstatix_0.6.0             readxl_1.3.1             
[29] foreign_0.8-80            carData_3.0-4            
[31] purrr_0.3.4               car_3.0-9                
[33] backports_1.1.8           ellipsis_0.3.1           
[35] abind_1.4-5               colorspace_1.4-1         
[37] ggsignif_0.6.0            stringi_1.4.6            
[39] munsell_0.5.0             broom_0.7.0              
[41] crayon_1.3.4  

#### what do you get with pkgbuild::check_build_tools()?
> pkgbuild::check_build_tools()
Your system is ready to build packages!
>

#### Is this a network drive?
No.

#### Are you working with a network drive at all?
No. 

#### Are there spaces in the path to R?
Only the space in Programe Files as here: "C:\Program Files\R\R-4.0.2\bin\x64"
This has not caused problem before though. 

#### Does this work: tmp <- tempfile(); cat(letters,file = tmp); readLines(tmp) (can you write to tempdir())?'
This is the output. The accent in my name is the one causing the encoding problem in the path below. I've been using mrgsolve on this computer with earlier verions of R before so not sure if it indicates any problem:
> tmp <- tempfile(); cat(letters,file = tmp); readLines(tmp)
[1] "a b c d e f g h i j k l m n o p q r s t u v w x y z"
Warning message:
In readLines(tmp) :
  incomplete final line found on 'C:\Users\mÃ¡s\AppData\Local\Temp\RtmpOOBqqi\file4841dc859e5'
> 

#### This: pkgbuild::check_compiler(debug = TRUE)?
This is the output:
> pkgbuild::check_compiler(debug = TRUE)
Trying to compile a simple C file
> Running "C:/PROGRA\~1/R/R-4.0.2/bin/x64/Rcmd.exe" SHLIB foo.c
> "C:/rtools40/mingw64/bin/"gcc  -I"C:/PROGRA\~1/R/R-4.0.2/include" -DNDEBUG          -O2 -Wall  -std=gnu99 -mfpmath=sse -msse2 -mstackrealign -c foo.c -o foo.o
> C:/rtools40/mingw64/bin/gcc -shared -s -static-libgcc -o foo.dll tmp.def foo.o -LC:/PROGRA\~1/R/R-4.0.2/bin/x64 -lR
[1] TRUE
>

**Comment by soutomas** [comment]
I've been using another machine for now while trying to solve this. I've checked through the outputs above for quite a few times now and the output from `debugonce` seems to be clearer now and I've updated that above. I just can't figure out what's issue though. Did I miss anything obvious above?

**Comment by kylebaron** [comment]
Thanks, @tmscs ; I'm sorry I missed your output from 2 weeks ago;  I'll look at it this morning after a couple of meetings and get back to you.  It does look like most of the stuff is in place and working.

**Comment by kylebaron** [comment]
I wonder if this is the culprit: ` "C:\Users\Tomás\AppData\Local\Temp\RtmpmGKSPx"`. 

Can you try this: 
Create a directory: `C:/project`

then 

```r
mod <- mcode_cache("m23", s23code, project = "C:/project")
```

I'd like to know what happens there.  

Additionally:

Create a directory: `C:/build`

then 

```r
mod <- mcode_cache("m23", s23code, project = "C:/project", soloc = "C:/build")
```

**Comment by kylebaron** [comment]
https://github.com/Rdatatable/data.table/issues/3078

**Comment by kylebaron** [comment]
```r
mod <- mcode_cache("m23", s23code, project = enc2native(tempdir()))
```

```r
tmp <- enc2native(tempdir())
mod <- mcode_cache("m23", s23code, project = tmp, soloc = tmp)
```

**Comment by soutomas** [comment]
Thanks so much Kyle, it works now! The minimum change for things to work is by specifying "soloc":
> mod <- mcode_cache("m23", s23code, soloc = "C:/build")

With the above, it works even without specifying the project folder. It did not work when I only specified the project folder without "soloc". So it looks like "soloc" here is the issue. 

This latter code here also did not work with the same error: 
> tmp <- enc2native(tempdir())
mod <- mcode_cache("m23", s23code, project = tmp, soloc = tmp)

May I ask if you know what is the reason for this error, so I can avoid this in the future?

**Comment by kylebaron** [comment]
Thanks for finding the minimal change.  That's really helpful.  

`soloc` is the place where mrgsolve compiles the model code; it needs to write out some files and compile.  It might also need to unload or remove a model dll file.  It does this inside `tempdir()`, but inside a couple of subfolders specific to the model name and platform you are running on. 

When you use `mread_cache`, it just writes the model file (what is in `s23code`) to that directory (here `tempdir()`)

So I wondered if this was some problem with the path to tempdir() but it looks like that isn't it (since `tempdir()` works fine for `project`). 

I guess it could be some permissions issue, but all of those directories for `soloc` should be getting created by you so there shouldn't be a reason why you couldn't access them.  

# Short term fix

Just keep using this.  You can put the following 
```r
options(mrgsolve.soloc = "C:/build")
```
at the top of your script or in `.Rprofile` in your `HOME` directory. 

But obviously I'd like to figure out why this is happening and fix the root cause so you don't always have to do this. I can throw out some ideas but I'm afraid it will be tedious. Would you be willing to so a screen share some time and try some things together?  I think we could make a lot of ground pretty quick.  If you'd be ok with that, could you email me at the address in my github profile?

Thanks,
Kyle

**Comment by soutomas** [comment]
Yes, we can look into this when you have a moment. Feel free to let me know how I can help.
